opsmngr/users.go

```diff
@@ -0,0 +1,245 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package opsmngr
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	atlas "go.mongodb.org/atlas/mongodbatlas"
+)
+
+const (
+	usersBasePath = "users"
+	projectUsersBasePath = "groups/%s/users"
+	orgUsersBasePath = "orgs/%s/users"
+)
+
+// UsersService provides access to the user related functions in the Ops Manager API.
+//
+// See more: https://docs.opsmanager.mongodb.com/current/reference/api/groups/
+type UsersService interface {
+	ListProjectUsers(context.Context, string, *atlas.ListOptions) ([]*User, *atlas.Response, error)
+	ListOrgUsers(context.Context, string, *atlas.ListOptions) ([]*User, *atlas.Response, error)
+	Get(context.Context, string) (*User, *atlas.Response, error)
+	GetByName(context.Context, string) (*User, *atlas.Response, error)
+	Create(context.Context, *User) (*User, *atlas.Response, error)
+	Delete(context.Context, string) (*atlas.Response, error)
+	RemoveFromProject(context.Context, string, string) (*atlas.Response, error)
+}
+
+// UsersServiceOp provides an implementation of the UsersService interface
+type UsersServiceOp service
+
+var _ UsersService = &UsersServiceOp{}
+
+// User wrapper for a user response, augmented with a few extra fields
+type User struct {
+	Username     string        `json:"username"`
+	Password     string        `json:"password,omitempty"`
+	FirstName    string        `json:"firstName,omitempty"`
+	LastName     string        `json:"lastName,omitempty"`
+	EmailAddress string        `json:"emailAddress,omitempty"`
+	ID           string        `json:"id,omitempty"`
+	Links        []*atlas.Link `json:"links,omitempty"`
+	Roles        []*UserRole   `json:"roles,omitempty"`
+}
+
+// UserRole denotes a single user role
+type UserRole struct {
+	RoleName string `json:"roleName"`
+	GroupID  string `json:"groupId,omitempty"`
+	OrgID    string `json:"orgId,omitempty"`
+}
+
+// UsersResponse represents a array of users
+type UsersResponse struct {
+	Links      []*atlas.Link `json:"links"`
+	Results    []*User    `json:"results"`
+	TotalCount int           `json:"totalCount"`
+}
+
+// ListProjectUsers gets all users in a project.
+//
+// See more: https://docs.opsmanager.mongodb.com/current/reference/api/groups/get-all-users-in-one-group/
+func (s *UsersServiceOp) ListProjectUsers(ctx context.Context, projectID string, opts *atlas.ListOptions) ([]*User, *atlas.Response, error) {
+	path := fmt.Sprintf(projectUsersBasePath, projectID)
```

Comment:
[q] I can't remember off the top of my head - are there any checks outside of this method to ensure it is only called for the OM context or will give an appropriate not supported error message?